pkg/scheduler/framework/plugins/podtopologyspread/plugin.go

```diff
@@ -141,9 +145,128 @@ func (pl *PodTopologySpread) EventsToRegister() []framework.ClusterEventWithHint
+	}
+
+	return framework.QueueSkip, nil
+}
+
+// getConstraints extracts topologySpreadConstraint(s) from the Pod spec.
+// If the Pod doesn't have any topologySpreadConstraint, it returns default constraints.
+func (pl *PodTopologySpread) getConstraints(pod *v1.Pod) ([]topologySpreadConstraint, error) {
+	var constraints []topologySpreadConstraint
+	var err error
+	if len(pod.Spec.TopologySpreadConstraints) > 0 {
+		// We have feature gating in APIServer to strip the spec
+		// so don't need to re-check feature gate, just check length of Constraints.
+		constraints, err = pl.filterTopologySpreadConstraints(
+			pod.Spec.TopologySpreadConstraints,
+			pod.Labels,
+			v1.DoNotSchedule,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("obtaining pod's hard topology spread constraints: %w", err)
+		}
+	} else {
+		constraints, err = pl.buildDefaultConstraints(pod, v1.DoNotSchedule)
+		if err != nil {
+			return nil, fmt.Errorf("setting default hard topology spread constraints: %w", err)
+		}
+	}
+	return constraints, nil
+}
+
+func (pl *PodTopologySpread) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
+	originalNode, modifiedNode, err := util.As[*v1.Node](oldObj, newObj)
+	if err != nil {
+		logger.Error(err, "unexpected objects in isSchedulableAfterNodeChange", "oldObj", oldObj, "newObj", newObj)
+		return framework.Queue, err
+	}
+
+	constraints, err := pl.getConstraints(pod)
+	if err != nil {
+		logger.Error(err, "get constraints from pod failed", "pod", klog.KObj(pod))
+		return framework.Queue, err
+	}
+
+	// framework.UpdateNodeLabel: return Queue when topologyKey is added/changed/removed in node's labels, else return QueueSkip.
+	if originalNode != nil && modifiedNode != nil {
+		for _, c := range constraints {
+			if originalNode.Labels[c.TopologyKey] != modifiedNode.Labels[c.TopologyKey] {
+				logger.V(5).Info("the topology key in the node was updated", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode), "topologyKey", c.TopologyKey)
+				return framework.Queue, nil
+			}
+		}
+		logger.V(5).Info("the node's label is updated, but not related to the Pod's topologySpreadConstraints", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+		return framework.QueueSkip, nil
+	}
+
+	// framework.Add: return Queue when node has topologyKey in its labels, else return QueueSkip.
+	if modifiedNode != nil {
+		if !nodeLabelsMatchSpreadConstraints(modifiedNode.Labels, constraints) {
+			logger.V(5).Info("new created node doesn't match pod topologySpreadConstraints", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+			return framework.QueueSkip, nil
+		}
+		logger.V(5).Info("node that match topologySpreadConstraints was created", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+		return framework.Queue, nil
+	}
+
+	// framework.Delete: return Queue when node has topologyKey in its labels, else return QueueSkip.
+	if originalNode != nil {
```

Comment:
No need to check `if originalNode != nil`, right? (because if it reaches here, it should be the deletion event)